src/library/scala/collection/StrictOptimizedIterableOps.scala

```diff
@@ -237,4 +237,10 @@ trait StrictOptimizedIterableOps[+A, +CC[_], +C]
     (l.result(), r.result())
   }
 
+  // Optimization avoids creation of second collection
+  override def tapEach[U](f: A => U): C  = {
+    foreach(f)
+    coll
```

Comment:
can't return `this`?

Reply:
Unfortunately, `StrictOptimizedSeqOps[..., C]` does not conform to `C`. I could play around with introducing a self-type to it, `self: C => ` and see if there are any issues.

Reply:
Ok, It seems that this solution does solve it, as long as you also enforce the `self: C =>` type on all extending `StrictOptimized{Seq,Set,Map,SortedMap, SortedSet}Ops`. I'd like to get @julienrf 's opinion though if he forsees any issues with that constraint, maybe it's intended to be extendable by non-`C` things too

Reply:
The exceptions used to be `StringOps` and `ArrayOps`. All regular collection types extend their `C` parameter.  Can we make any further simplifications if we always enforce this? I don't think it's worth making this change specifically for `tapEach`. There are plenty of other methods that return `C`. In most cases you abstract over consuming a collection where this is not an issue. If you also need to produce a more specific collection type you need to use the proper abstraction anyway (unless we can change it in a more generic way).

Reply:
@szeiger I can't think of any other improvements or simplifications that could be made if it had `{ self: C => ` unfortunately.